java/arcs/core/host/ArcHostContextParticle.kt

```diff
@@ -84,57 +96,36 @@ class ArcHostContextParticle : AbstractArcHostParticle() {
      * is stored in de-normalized format.
      */
     suspend fun readArcHostContext(
-        arcHostContext: ArcHostContext,
-        hostId: String,
-        instantiateParticle: suspend (ParticleIdentifier) -> Particle
+        arcHostContext: ArcHostContext
     ): ArcHostContext? {
         val arcId = arcHostContext.arcId
 
         try {
-            val arcStateEntity = handles.arcHostContext.fetch()
-            val particleEntities = handles.particles.fetchAll()
-            val connectionEntities = handles.handleConnections.fetchAll()
-
-            // This arc has never been serialized before
-            if (arcStateEntity == null ||
-                particleEntities.isEmpty() ||
-                connectionEntities.isEmpty()) {
-                return null
-            }
+            // TODO(cromwellian): replace with .query(arcId, hostId) when queryHandles are efficient
+            val arcStateEntity = handles.arcHostContext.fetch() ?: return null
+            val particleEntities = arcStateEntity.particles.map { it.dereference() }.toSet()
 
-            // construct a map of particleName to Particle instance
-            val instantiatedParticles = particleEntities.map {
-                it.particleName to instantiateParticle(ParticleIdentifier.from(it.location))
-            }.associateBy({ it.first }, { it.second })
+            val particles = particleEntities.map {
+                requireNotNull(it) {
+                    "Particle dereference was null for $arcId"
```

Comment:
Small suggestion: you could move the requireNotNull up into the previous `map`, and then you'd be able to give a more useful error message (e.g. "Encountered invalid particle reference when deserialising ArcHost" or something like that)

Also, you could delete the `particleEntities` variable and just chain these:

```
val particles = arcStateEntity.particles
    .map { ... }
    .toSet()
    .map { ... }
```